README.md
# bitjack21 - radiojack.one (radiocoin blackjack)
* could be a docker image but its not
 
* changed chips from a floating decimal only to a round number as a balance or playable balance or withdraw-able balance
* to do, change 1,5,10 chips to 10,50,100 via png edit
# -> done * readded deposit by QR code 
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/qr-deposit.png)

bitjack21.com
```
code has pre-existing graphs in graph_profits.php
code uses pre-existing crude admin.php panel 
must change if user != c4pt to admin username for site

admin.php:if(!isset($_SESSION['username']) || ($_SESSION['username'] != 'c4pt' && $_SESSION['username'] != 'c4pt'))
admin.php:  $q1 = mysqli_query($con, "select sum(netGain)/-100 as gain from games where player != 'c4pt'");
```

game play logic and min / max bet in control.php
```
define('MAX_BET', 1000);
define('MIN_BET', 100);

```

requires radiocoin, mysql, httpd, phpmyadmin (or managing the database manually with mysql)
* https://github.com/c4pt000/radiocoin/releases/download/linux/radiocoin-6.0.1_radiocoin-2.x86_64.rpm

```
  yum install nano sudo wget git-core httpd ufw mysql-server mysql phpMyAdmin.noarch -y
  yum install radio*rpm -y
  rpm -e --nodeps firewalld firewalld-filesystem
  echo "IPV6=no" >> /etc/ufw/ufw.conf
  systemctl enable ufw
  ufw enable
  ufw allow in 80
  ufw allow out 53
  ufw allow out 80
  ufw allow out 443
  ufw allow out 9332
  ufw allow out 9333
  ufw allow out 3306
  ufw status verbose
  ufw status numbered
  ufw show listening
```

disable phpmyadmin (if not in use or safe guarded) will be visible and accessible from outside WAN
```
phpMyAdmin.conf goes in /etc/httpd/conf.d/
welcome.conf goes in /etc/httpd/conf.d/
radiocoin.conf goes in /root/.radiocoin/radiocoin.conf
fix-perm.sh to set permissions for /var/www/html
setenforce 0 (disable selinux, at your own risk)
```

```
install mariadb (formerly mysql) + httpd (and or phpmyadmin)
 systemctl start mariadb
 systemctl enable mariadb

systemctl enable httpd
systemctl start httpd
copy * to /var/www/html
sh fix-perm.sh
systemctl restart httpd
mysql_secure_installation 
create database "radiojack" or change database name accordingly in common.php
install database template
mysql -u root -p radiojack < bitjack21_DB.sql
```
navigate to 127.0.0.1/phpMyAdmin            where 127.0.0.1 is your WAN ip

use your own security guidelines this is just a template
set mysql connection in common.php change yourpassword-here to the root password (or user password) that can access "radiojack"
```
function connectDB()
{

$sql = new mysqli('localhost', 'root', 'yourpassword-here', 'radiojack');
 
if($sql->connect_errno > 0){
    die('Unable to connect to database [' . $sql->connect_error . ']');
}
 
$con = mysqli_connect("localhost","root","yourpassword-here","radiojack") or die("Some error occurred during connection " . mysqli_error($con));


  return $con;


}
```

also in common.php the radiocoin.conf connection (must match radiocoin.conf rpcuser/rpcpassword)
rpcallowip=127.0.0.1 in radiocoin.conf only allows connections from your local LAN (no WAN access) to safe guard rpc commands remotely
must also keep paytxfee=1.00 for a minimum to remove from the site wallet balance in order to be able to send
```
function connectBitcoin()
{


          $bitcoin = new jsonRPCClient('http://root:rpcpassword@127.0.0.1:9332/');

	  return $bitcoin;
}
```

![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-deposit.png)
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-deposit-detect1.png)
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-regular-withdraw.png)
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-sent-manual-withdrawl.png)
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-in-game-1.png)
![s1](https://raw.githubusercontent.com/c4pt000/bitjack21-radiojack.one-radiocoin-blackjack/master/radiojack-in-game-2.png)


history.txt
    1  yum install nano sudo wget git-core httpd
    2  cd /opt/
    3  git clone https://github.com/c4pt000/bitjack21
    4  ls
    5  cd bitjack21/
    6  ls
    7  systemctl enable httpd
    8  systemctl start httpd
    9  ls
   10  ifconfig 
   11  systemctl start httpd
   12  rpm -e --nodeps firewalld firewalld-filesystem
   13  yum install ufw -y
   14  ufw enable
   15  ufw allow out 53
   16  ufw allow out 80
   17  ufw allow out 443
   18  nano /etc/ufw/ufw.conf 
   19  ufw enable
   20  ufw status verbose
   21  ufw delete 21
   22  ufw delete 2
   23  ufw status verbose
   24  ufw allow in 80
   25  nano /etc/httpd/conf.d/welcome.conf 
   26  systemctl restart httpd
   27  yum install mariadb-server mariadb-client -y
   28  yum install mysql-server mysql
   29  yum search phpmy
   30  yum install phpMyAdmin.noarch
   31  cd /
   32  wget https://github.com/c4pt000/radiocoin/releases/download/linux/radiocoin-6.0.1_radiocoin-2.x86_64.rpm
   33  yum install radio*rpm -y
   34  mkdir /root/.radiocoin
   35  nano /root/.radiocoin/radiocoin.conf
   36  radiocoind 
   37  radiocoin-cli getinfo
   38  cd /opt/bitjack21/
   39  ls
   40  cp -rf * /var/www/html/
   41  ls -lh /var/www/html/
   42  restorecon -Rv * /var/www/html/
   43  chmod -R 755 /var/www/html/
   44  systemctl restart httpd
   45  cd /var/www/html/
   46  ls
   47  nano index.php 
   48  grep -ri "https://"
   49  find . -type f -print0 | xargs -0 sed -i -e 's/bitjack21.com/radiojack.one/g' 
   50  systemctl restart httpd
   51  grep -ri "https"
   52  grep -ri "https:"
   53  find . -type f -print0 | xargs -0 sed -i -e 's/https:/http:/g' 
   54  systemctl restart httpd
   55  cp -rf index.php index.php.orig
   56  nano index.php
   57  systemctl restart httpd
   58  nano index.php
   59  systemctl restart httpd
   60  cd /var/www/html/
   61  nano index.php
   62  ls
   63  nano register.php 
   64  nano index.php
   65  nano /etc/httpd/conf.d/phpMyAdmin.conf 
   66  chmod -R 755 /usr/share/phpMyAdmin
   67  restorecon -Rv /usr/share/phpMyAdmin
   68  nano /etc/httpd/conf.d/phpMyAdmin.conf 
   69  systemctl restart httpd
   70  nano /etc/httpd/conf.d/phpMyAdmin.conf 
   71  systemctl restart httpd
   72  systemctl status httpd
   73  nano /etc/httpd/conf.d/phpMyAdmin.conf 
   74  systemctl restart httpd
   75  nano /etc/httpd/conf.d/phpMyAdmin.conf 
   76  systemctl restart httpd
   77  mysql_secure_installation 
   78  systemctl start mysqld
   79  systemctl start mysql
   80  systemctl start mariadb
   81  systemctl enable mariadb
   82  cd /var/www/html/
   83  cat /var/log/php-fpm/www-error.log 
   84  grep -ri getBTCBalance
   85  grep -ri getRADCBalance
   86  nano index.php
   87  nano common.php 
   88  nano withdraw.php 
   89  grep -ri "0.01"
   90  cat /var/log/php-fpm/www-error.log 
   91  date
   92  cat /var/log/php-fpm/www-error.log 
   93  nano index.php
   94  cat /var/log/php-fpm/www-error.log 
   95  date
   96  nano withdraw.php 
   97  nano login.php 
   98  nano index.php
   99  nano withdraw.php 
  100  cat /var/log/php-fpm/www-error.log 
  101  nano wd.php 
  102  cat /var/log/php-fpm/www-error.log 
  103  nano wd.php 
  104  nano /root/.radiocoin/radiocoin.conf 
  105  radiocoin-cli stop
  106  radiocoind
  107  radiocoin-cli getinfo
  108  radiocoin-cli getbalance
  109  cat /var/log/php-fpm/www-error.log 
  110  nano wd.php 
  111  cat /var/log/php-fpm/www-error.log 
  112  nano wd.php 
  113  cat /var/log/php-fpm/www-error.log 
  114  nano play.php 
  115  nano control.php 
  116  nano play.php 
  117  nano js/bitjack21.js 
  118  ls
  119  nano control.php 
  120  nano play.php 
  121  find . -name 'double.php'
  122  nano play.php 
  123  grep -ri "double.php"
  124  find . -name "double.php"
  125  nano play.php 
  126  nano control.php 
  127  cat /var/log/php-fpm/www-error.log 
  128  nano control.php 
  129  nano login.php 
  130  nano common.php 
  131  nano control.php 
  132  nano play.php 
  133  pwd
  134  ifconfig 
  135  nano js/bitjack21.js 
  136  cp -rf /bitjack21.js js/
  137  nano js/bitjack21.js 
  138  cat /var/log/php-fpm/www-error.log 
  139  date
  140  nano js/bitjack21.js 
  141  cp -rf control.php js/
  142  nano js/jquery-1.6.2.min.js 
  143  nano play.php 
  144  cat /var/log/php-fpm/www-error.log 
  145  nano control.php 
  146  cat /var/log/php-fpm/www-error.log 
  147  nano control.php 
  148  cat /var/log/php-fpm/www-error.log 
  149  nano control.php 
  150  cat /var/log/php-fpm/www-error.log 
  151  grep -ri "RADCRound()"
  152  grep -ri "BTCRound()"
  153  grep -ri "BTCRound"
  154  grep -ri "RADCRound"
  155  nano common.php 
  156  cat /var/log/php-fpm/www-error.log 
  157  nano common.php 
  158  nano control.php 
  159  grep -ri CHIPSPERBTC
  160  grep -ri CHIPSPERRADC
  161  nano common.php 
  162  nano index.php
  163  cat /var/log/php-fpm/www-error.log 
  164  nano wd.php 
  165  grep -ri "chips"
  166  nano common.php 
  167  radiocoin-cli getbalance
  168  radiocoin-cli sendtoaddress RX9oPtdwdWsaXomiQJg2QeU7ZLXDiqd5w4 102
  169  radiocoin-cli sendtoaddress RX9oPtdwdWsaXomiQJg2QeU7ZLXDiqd5w4 101
  170  radiocoin-cli getbalance
  171  nano play.php 
  172  nano index.php
  173  nano play.php 
  174  nano index.php
  175  nano play.php 
  176  rm -rf play.php 
  177  wget https://raw.githubusercontent.com/c4pt000/bitjack21/master/play.php
  178  nano css/layout.css 
  179  nano css/style.css 
  180  nano play.php 
  181  cp -rf play.php play.php.orig
  182  nano play.php
  183  nano css/layout.css 
  184  nano css/style.css 
  185  nano play.php
  186  nano rules.php 
  187  radiocoin-cli getinfo
  188  radiocoin-cli getbalance
  189  grep -ri "exceeds"
  190  nano withdraw.php 
  191  grep -ri "data.error"
  192  nano js/bitjack21.MAX.js 
  193  nano withdraw.php 
  194  nano css/layout.css 
  195  nano css/style.css 
  196  nano css/layout.css 
  197  nano css/style.css 
  198  nano play.php
  199  nano common.php 
  200  nano play.php

common.php
<?php

require_once "jsonRPCClient.php";

define('CHIPSPERBTC', 1);
$sessionsalt = 'rakdsayratdsa';

$mail_headers = 'X-Mailer: RadioJack.one' . "\r\n" .
    'Reply-To: radiojack@gmail.com' . "\r\n" .
    'Errors-To: radiojack@gmail.com';
// ADMIN: Put your email address in the 2 lines above


function mail_payout($email, $user, $amount, $wallet, $txid)
{
   global $mail_headers;

   $message = "Hello,\n\nThis is a RADC withdrawal notification from http://www.radiojack.one for $user.\n\n$amount RADC has just been sent to your Bitcoin wallet $wallet\n\nYou can track this payout on blockexplorer.com using the transaction ID $txid as per below:\n\nhttp://blockexplorer.com/tx/$txid\n\nThanks for playing!\n\n--\nradiojack.one";

   mail($email, "[RadioJack] Withdrawal Notification", $message, $mail_headers);
}

function mail_manual_payout($email, $user, $amount, $wallet)
{
   global $mail_headers;

   $message = "Hello,\n\nThis is a notice that you have a pending RADC withdrawal from http://www.radiojack.one for $user.\n\nA withdrawal of $amount RADC is now queued for manual processing and will be sent within 24 hours to address $wallet\n\nThanks for playing!\n\n--\nradiojack.one";

   $time = date("Y-m-d H:i:s +0000");

   mail($email, "[RadioJack] Pending Withdrawal Notification", $message, $mail_headers);
   mail('ADMIN_EMAIL_ADDY@gmail.com', '[RadioJack] MANUAL PAYOUT REQUEST', "USER=$user\nAMOUNT=$amount\nWALLET=$wallet\nTIME=$time", $mail_headers);
}

function mail_error($msg)
{
   global $mail_headers;
   mail('radiojack@gmail.com', '[RadioJack] ERROR', $msg, $mail_headers);
}

function drawmenu()
{
  echo '<ul id="a">';
  if(isset($_SESSION['username']))
  {
    echo <<< DONE
    <li class="mb1"><a href="index.php">Home</a></li>
    <li class="mb1"><a href="play.php">Play</a></li>
    <li class="mb1"><a href="withdraw.php">Deposit/Withdraw</a></li>
    <li class="mb1"><a href="rules.php">Rules</a></li>
    <li class="mb1"><a href="contact.php">Contact</a></li>
    <li class="mb1"><a href="logout.php">Log Out</a></li>
DONE;
  }
  else
  {
    echo <<< DONE
    <li class="mb2"><a href="index.php">Home</a></li>
    <li class="mb2"><a href="register.php">Register</a></li>
    <li class="mb2"><a href="rules.php">Rules</a></li>
    <li class="mb2"><a href="contact.php">Contact</a></li>
DONE;
  }
  echo '</ul>';
}


function hash_password($password, $nonce) {
  $site_key = '12qwerty56';
  return hash_hmac('sha512', $password . $nonce, $site_key);
}

function generateSalt($max = 15) {
	$characterList = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
	$i = 0;
	$salt = "";
	do {
		$salt .= $characterList{mt_rand(0,strlen($characterList)-1)};
		$i++;
	} while ($i <= $max);
	return $salt;
}


function connectBitcoin()
{


          $bitcoin = new jsonRPCClient('http://root:rpcpassword@127.0.0.1:9332/');

	  return $bitcoin;
}

function connectDB()
{

$sql = new mysqli('localhost', 'root', 'yourpassword-here', 'radiojack');
 
if($sql->connect_errno > 0){
    die('Unable to connect to database [' . $sql->connect_error . ']');
}
 
$con = mysqli_connect("localhost","root","yourpassword-here","radiojack") or die("Some error occurred during connection " . mysqli_error($con));


  return $con;


}

function getBTCBalance($user=null)
{
  if(is_null($user) && !isset($_SESSION['username']))
  {
    return 0;
  }
  if(is_null($user))
  {
    $user = $_SESSION['username'];
  }
  $bal = null;
  $bitcoin = connectBitcoin();
  try {
    $bal = ($bitcoin->getbalance($user));
  } catch (Exception $e) {
    return null;
  }
  return BTCRound($bal);
}


function BTCRound($value)
{
//  return floor(($value) * 1e8 + .5) * .00000001;
//  return floor(($value) * 1e8 + .5) * .00000001;
return number_format((float)$value, 0, '.', '');
}

function chipRound($value)
{
//    return round($value * 1e8);
//  return floor(($value) * 1e6 + .5) * .000001;
//  return round(($value) * 1e6 ) * .000001;
return number_format((float)$value, 0, '.', '');

}

function getBalance()
{
  $bitcoin = connectBitcoin();
  $bal = null;
  try {
    $bal = ($bitcoin->getbalance($_SESSION['username']))*CHIPSPERBTC;
  } catch (Exception $e) {
    return null;
  }
  return chipRound($bal);
}

function getFlooredBalance()
{
  return /*floor*/(getBalance());
//  return floor(getBalance();
}

/***********************************************************************************/

function SHA512_encode($str)
{
  return base64_encode(bin2hex(hash('sha512',$str)));
}

function validate_session()
{
    global $sessionsalt;

    if(!isset($_SESSION)) {
    session_start();}

    if (!isset($_SESSION['SERVER_GENERATED_SID']))
    {
      reset_session();
    }

    if(isset($_SESSION['G5X27CIAKGB']) && $_SESSION['G5X27CIAKGB'] !== SHA512_encode($_SERVER['REMOTE_ADDR'].$_SERVER['HTTP_USER_AGENT'].$sessionsalt))
    {
      reset_session();
    }

    ++$_SESSION['loads'];

    if(isset($_SESSION['username']))
    {
	$u = $_SESSION['username'];
	$s = session_id();

	$con = connectDB();
	if(!$con)
	{
	    reset_session();
	}
	else
	{
	    $ok = false;

	    if ($query = mysqli_query($con, "SELECT EXISTS(SELECT 1 FROM users WHERE username = '$u' AND session = '$s' LIMIT 1)"))
	    {    
		$result = mysqli_fetch_row($query);
		$ok =  (bool)($result[0]);
	    }
	    else
	    {
		$ok =  false;
	    }
	  
	    if(!$ok)
	    {
		reset_session();
	    }
	    else
	    {
		mysqli_query($con, "UPDATE users set lastactive=NOW() where username='$u'");
	    }
	}
    }
}
/******************************************************************************/



function allow_session()
{
  if(!isset($_SESSION['username']))
  {
    return false;
  }
  $user = $_SESSION['username'];
  $ses  = session_id();

  $con = connectDB();
  if(!$con)
  {
    return false;
  }

  $result = mysqli_query($con, "UPDATE users set session='$ses' where username='$user'");
   if(!$result) { return false; }

  return true;
}

function reset_session()
{
global $sessionsalt;
 /* if(isset($_COOKIE[session_name()]))
  {
    // Kill the cookie assocated to the session.
    if (ini_get("session.use_cookies"))
    {
        $params = session_get_cookie_params();
        setcookie(session_name(), '', time() - 42000,
            $params["path"], $params["domain"],
            $params["secure"], $params["httponly"]);
    }
  }

  session_start();
   //Destroy the session itself.
   session_regenerate_id();
   //session_destroy();
   unset($_SESSION);
   $_SESSION = array();
session_destroy();
   session_start();*/


if(!isset($_SESSION)) {
session_start();}
//session_regenerate_id();
//session_destroy();
unset($_SESSION);
session_destroy();
session_start();
session_regenerate_id();
   $_SESSION['SERVER_GENERATED_SID'] = true;

  $_SESSION['G5X27CIAKGB'] = SHA512_encode($_SERVER['REMOTE_ADDR'].$_SERVER['HTTP_USER_AGENT'].$sessionsalt);

  // Set number of loads to 0 in the session
  $_SESSION['loads'] = 0;


}


?>
